crypto/asn1/a_i2d_fp.c

```diff
@@ -61,6 +61,43 @@
 #include <openssl/mem.h>
 
 
+int ASN1_i2d_bio(i2d_of_void *i2d, BIO *out, void *in) {
+  if (i2d == NULL || out == NULL || in == NULL) {
+    OPENSSL_PUT_ERROR(ASN1, ERR_R_PASSED_NULL_PARAMETER);
+    return 0;
+  }
+
+  int size = i2d(in, NULL);
+  if (size <= 0) {
+    return 0;
+  }
+
+  char *buffer = (char *)OPENSSL_malloc(size);
+  if (buffer == NULL) {
+    OPENSSL_PUT_ERROR(ASN1, ERR_R_MALLOC_FAILURE);
+    return 0;
+  }
+
+  unsigned char *outp = (unsigned char *)buffer;
+  i2d(in, &outp);
+
+  int written = 0, idx=0;
+  for (;;) {
+    written = BIO_write(out, &(buffer[idx]), size);
+    if (written == size) {
+      break;
+    }
+    if (written <= 0) {
```

Comment:
* Should we call `OPENSSL_PUT_ERROR` here?
* `BIO_write` returning `0` is not an error, it just means that no bytes were written. How should we handle that?

Reply:
* Added
* Even though returning 0 isn't an actual error from `BIO_write`, this still means that the bytes weren't all properly written to `out`. The previous block of code checks if all `size` bytes was written to `bio`.

Reply:
Consolidating this block to use `BIO_write_all` instead, like much of the other related APIs are using.